pkg/threescale/connectors/backend/cached_backend.go

```diff
@@ -0,0 +1,261 @@
+type CacheValue map[string]map[backend.LimitPeriod]*Limit
+
+// CachedBackend provides a pluggable cache enabled 'Backend' implementation
+// Supports reporting metrics for non-cached responses.
+type CachedBackend struct {
+	ReportFn      metrics.ReportMetricsFn
+	cache         Cacheable
+	mutex         sync.RWMutex
+	hierarchyTree hierarchyTree
+}
+
+type LocalCache struct {
+	ds cmap.ConcurrentMap
+}
+
+// map to represent parent --> children metric relationship
+type metricParentToChildren map[string][]string
+
+// hierarchyTree - represents the relationship between service metrics and their children
+type hierarchyTree map[string]metricParentToChildren
+
+// NewCachedBackend returns a pointer to a CachedBackend with a default LocalCache if no custom implementation has been provided
+func NewCachedBackend(cache Cacheable, rfn metrics.ReportMetricsFn) *CachedBackend {
+	if cache == nil {
+		cache = NewLocalCache()
+	}
+	return &CachedBackend{
+		ReportFn:      rfn,
+		cache:         cache,
+		hierarchyTree: make(hierarchyTree),
+	}
+}
+
+// NewLocalCache returns a pointer to a LocalCache with an initialised empty data structure
+func NewLocalCache() *LocalCache {
+	return &LocalCache{ds: cmap.New()}
+}
+
+// Get entries for LocalCache
+func (l LocalCache) Get(cacheKey string) (CacheValue, bool) {
+	var cv CacheValue
+	v, ok := l.ds.Get(cacheKey)
+	if !ok {
+		return cv, ok
+	}
+
+	cv = v.(CacheValue)
+	return cv, ok
+}
+
+// Set entries for LocalCache
+func (l LocalCache) Set(cacheKey string, cv CacheValue) {
+	l.ds.Set(cacheKey, cv)
+}
+
+// AuthRep provides a combination of authorizing a request and reporting metrics to 3scale
+func (cb CachedBackend) AuthRep(req AuthRepRequest, c *backend.ThreeScaleClient) (Response, error) {
+	// compute the cache key for this request
+	hierarchyKey := generateHierarchyKey(req.ServiceID, c.GetPeer())
+	appIdentifier := generateAppIdentifier(req)
+	cacheKey := generateCacheKey(hierarchyKey, appIdentifier)
+
+	var affectedMetrics backend.Metrics
+	// gather the metrics affected by this request using the hierarchy information we might have gathered
+	affectedMetrics, ok := cb.computeAffectedMetrics(hierarchyKey, req.Params.Metrics)
+
+	// instrumentation
+	mc := metricsConfig{ReportFn: cb.ReportFn, Endpoint: "AuthRep", Target: "Backend"}
+
+	if !ok {
+		// !! CACHE MISS - we don't know about this service id and host combination!!
+		// we need to build hierarchy map blanket call to authRep with extensions enabled and no usage reported
+		resp, err := callRemote(req, map[string]string{"hierarchy": "1"}, c, mc)
+		if err != nil {
+			fmt.Println("error calling 3scale with blanket request")
+			// need to do some handling here and potentially mark this hierarchyKey as fatal to avoid calling 3scale with repeated invalid/failing requests
+			return Response{}, err
+		}
+		// populate the hierarchy tree
+		cb.setTreeEntry(hierarchyKey, resp.GetHierarchy())
```

Comment:
I think from the point of view of the caller of this function, this is infallible, it could perhaps be an empty map which would be fine.

https://github.com/3scale/3scale-go-client/blob/e366adc214e96578b5e53fd5b0250c5cfaec414c/client/client.go#L294-L307

Reply:
Yes, I am considering the case where for whatever reason the response does not contain a well formed response for hierarchies and it fails to parse.

Reply:
if the client fails to decode the xml then the caller gets an error anyway. I think its safe but if you have any suggestions looking at the client code I will be happy to make the change there for sure.